# AUTONORMALIZE DEMO 
Using Autonormalize to normalize a kaggle dataset about food purchasing.

In [1]:
import os

import pandas as pd
import autonormalize as an

In [2]:
food_df = pd.read_csv(os.path.join(os.getcwd(), 'autonormalize/downloads/food.csv'), encoding='latin1')
food_df = food_df.drop(columns=food_df.columns[10:])
print("Rows: "+ str(food_df.shape[0]))
print("Columns: " + str(food_df.shape[1]))
food_df.head(3)

Rows: 21477
Columns: 10


Area Abbreviation  Area Code         Area  Item Code  \
0               AFG          2  Afghanistan       2511   
1               AFG          2  Afghanistan       2805   
2               AFG          2  Afghanistan       2513   

                       Item  Element Code Element         Unit  latitude  \
0        Wheat and products          5142    Food  1000 tonnes     33.94   
1  Rice (Milled Equivalent)          5142    Food  1000 tonnes     33.94   
2       Barley and products          5521    Feed  1000 tonnes     33.94   

   longitude  
0      67.71  
1      67.71  
2      67.71

In [3]:
grps = food_df.groupby(['Area', 'Element'])


In [20]:
print(grps.indices)
for name, group in grps:
    print(group.indices)

{('Afghanistan', 'Feed'): array([ 2,  4,  9, 10, 14, 56, 60, 63, 65, 79]), ('Afghanistan', 'Food'): array([ 0,  1,  3,  5,  6,  7,  8, 11, 12, 13, 15, 16, 17, 18, 19, 20, 21,
       22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38,
       39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55,
       57, 58, 59, 61, 62, 64, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76,
       77, 78, 80, 81, 82]), ('Albania', 'Feed'): array([ 83,  85,  87,  89,  92,  94,  95,  96,  99, 103, 107, 110, 117,
       119, 130, 165, 169, 178, 180, 182, 184, 187, 190, 200, 202]), ('Albania', 'Food'): array([ 84,  86,  88,  90,  91,  93,  97,  98, 100, 101, 102, 104, 105,
       106, 108, 109, 111, 112, 113, 114, 115, 116, 118, 120, 121, 122,
       123, 124, 125, 126, 127, 128, 129, 131, 132, 133, 134, 135, 136,
       137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149,
       150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162,
       163, 164, 166, 167, 1

AttributeError: 'DataFrame' object has no attribute 'indices'

This dataset has 21477 rows and we've cut it down to 10 columns. As you can see, there are many data dependencies between that columns that obviously should be split up. For example, Area, Area Code, and Area Abreviation obviously should be dependent on each other.

In [4]:
deps_approx = an.find_dependencies(food_df, 0.96)

100%|██████████| 10/10 [00:02<00:00,  3.54it/s]


In [5]:
from autonormalize import normalize
depdf = normalize.DepDF(deps_approx, food_df)
normalize.normalize_dataframe(depdf)

['Area']
['latitude']


In [6]:
normalize.make_indexes(depdf)

In [5]:
print(depdf.deps)

 --> Area
 --> Item Code
 --> Item
 {Element}  --> Element Code
 {Element Code}  --> Element


In [10]:
print(depdf.children[0].deps)

 {Area Abbreviation}  {Area Code}  --> Area
 {Area}  {Area Code}  --> Area Abbreviation
 {Area}  {Area Abbreviation}  --> Area Code
 {Area}  {Area Abbreviation}  {Area Code}  --> latitude
 {Area}  {Area Abbreviation}  {Area Code}  --> longitude


In [11]:
print(depdf.children[0].children[0].deps)

 --> longitude
 {longitude}  --> Unit


In [4]:
groupings = an.normalize_dependencies(deps_approx)
for grp in groupings:
    print('\n~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~\n')
    print(grp)
    print(grp.get_prim_key())


~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~

 --> Area Code
 --> Item Code
 --> Item
 {Element}  --> Element Code
 {Element Code}  --> Element
['Item', 'Item Code', 'Element', 'Area Code']

~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~

 {Area Abbreviation}  {Area}  --> Area Code
 {Area}  {Area Code}  --> Area Abbreviation
 {Area Abbreviation}  {Area Code}  --> Area
 {Area Abbreviation}  {Area}  {Area Code}  --> latitude
 {Area Code}  {Area}  {Area Code}  --> longitude
['Area Code']

~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~

 --> latitude
 {latitude}  --> Unit
['latitude']


In [5]:
new_dfs = an.split_dataframe(food_df, groupings)

In [6]:
new_dfs[0]

Area Code  Item Code                          Item  Element Code  \
0              2       2511            Wheat and products          5142   
1              2       2805      Rice (Milled Equivalent)          5142   
2              2       2513           Barley and products          5521   
3              2       2513           Barley and products          5142   
4              2       2514            Maize and products          5521   
5              2       2514            Maize and products          5142   
6              2       2517           Millet and products          5142   
7              2       2520                Cereals, Other          5142   
8              2       2531         Potatoes and products          5142   
9              2       2536                    Sugar cane          5521   
10             2       2537                    Sugar beet          5521   
11             2       2542        Sugar (Raw Equivalent)          5142   
12             2       2543             Sweeteners, Other          5142   
13             2       2745                         Honey          5142   
14             2       2549    Pulses, Other and products          5521   
15             2       2549    Pulses, Other and products          5142   
16             2       2551             Nuts and products          5142   
17             2       2560         Coconuts - Incl Copra          5142   
18             2       2561                   Sesame seed          5142   
19             2       2563  Olives (including preserved)          5142   
20             2       2571                  Soyabean Oil          5142   
21             2       2572                 Groundnut Oil          5142   
22             2       2573             Sunflowerseed Oil          5142   
23             2       2574          Rape and Mustard Oil          5142   
24             2       2575                Cottonseed Oil          5142   
25             2       2577                      Palm Oil          5142   
26             2       2579                Sesameseed Oil          5142   
27             2       2580                     Olive Oil          5142   
28             2       2586           Oilcrops Oil, Other          5142   
29             2       2601         Tomatoes and products          5142   
...          ...        ...                           ...           ...   
21447        181       2765                   Crustaceans          5142   
21448        181       2766                   Cephalopods          5142   
21449        181       2767               Molluscs, Other          5142   
21450        181       2775                Aquatic Plants          5142   
21451        181       2680                   Infant food          5142   
21452        181       2905      Cereals - Excluding Beer          5521   
21453        181       2905      Cereals - Excluding Beer          5142   
21454        181       2907                 Starchy Roots          5142   
21455        181       2908                   Sugar Crops          5142   
21456        181       2909            Sugar & Sweeteners          5142   
21457        181       2911                        Pulses          5142   
21458        181       2912                      Treenuts          5142   
21459        181       2913                      Oilcrops          5521   
21460        181       2913                      Oilcrops          5142   
21461        181       2914                Vegetable Oils          5142   
21462        181       2918                    Vegetables          5142   
21463        181       2919       Fruits - Excluding Wine          5142   
21464        181       2922                    Stimulants          5142   
21465        181       2923                        Spices          5142   
21466        181       2924           Alcoholic Beverages          5142   
21467        181       2943                          Meat          5142   
21468        181       2945     

In [7]:
new_dfs[1]

Area Abbreviation  Area Code                                Area  \
0                 ARM          1                             Armenia   
1                 AFG          2                         Afghanistan   
2                 ALB          3                             Albania   
3                 DZA          4                             Algeria   
4                 AGO          7                              Angola   
5                 ATG          8                 Antigua and Barbuda   
6                 ARG          9                           Argentina   
7                 AUS         10                           Australia   
8                 AUT         11                             Austria   
9                 AZE         12                             Bahamas   
10                BRB         14                            Barbados   
11                BGD         16                          Bangladesh   
12                BMU         17                             Bermuda   
13                BOL         19    Bolivia (Plurinational State of)   
14                BWA         20                            Botswana   
15                BRA         21                              Brazil   
16                BLZ         23                              Belize   
17                SLB         25                     Solomon Islands   
18                BRN         26                   Brunei Darussalam   
19                BGR         27                            Bulgaria   
20                MMR         28                             Myanmar   
21                CMR         32                            Cameroon   
22                CAN         33                              Canada   
23                CPV         35                          Cabo Verde   
24                CAF         37            Central African Republic   
25                LKA         38                           Sri Lanka   
26                TCD         39                                Chad   
27                CHL         40                               Chile   
28                CHN         41                     China, mainland   
29                COL         44                            Colombia   
..                ...        ...                                 ...   
144               SWE        210                              Sweden   
145               CHE        211                         Switzerland   
146               TKM        213                        Turkmenistan   
147               CHN        214           China, Taiwan Province of   
148               TZA        215         United Republic of Tanzania   
149               THA        216                            Thailand   
150               TGO        217                                Togo   
151               TTO        220                 Trinidad and Tobago   
152               OMN        221                                Oman   
153               TUN        222                             Tunisia   
154               TUR        223                              Turkey   
155               ARE        225                United Arab Emirates   
156               UGA        226                              Uganda   
157               GBR        229                      United Kingdom   
158               UKR        230                             Ukraine   
159               USA        231            United States of America   
160               BFA        233                        Burkina Faso   
161               URY        234                             Uruguay   
162               UZB        235                          Uzbekistan   
163               VEN        236  Venezuela (Bolivarian Republic of)   
164               VNM        237                            Viet Nam   
165               ETH        238                            Ethiopia   
166               WSM        244                               Samoa   
167               YEM        249            

In [8]:
new_dfs[2]

Unit  latitude
0    1000 tonnes    -40.90
1    1000 tonnes    -38.42
2    1000 tonnes    -35.68
3    1000 tonnes    -32.52
4    1000 tonnes    -30.56
5    1000 tonnes    -29.61
6    1000 tonnes    -26.52
7    1000 tonnes    -25.27
8    1000 tonnes    -23.44
9    1000 tonnes    -22.96
10   1000 tonnes    -22.33
11   1000 tonnes    -20.90
12   1000 tonnes    -20.35
13   1000 tonnes    -19.02
14   1000 tonnes    -18.77
15   1000 tonnes    -18.67
16   1000 tonnes    -17.68
17   1000 tonnes    -16.58
18   1000 tonnes    -16.29
19   1000 tonnes    -15.38
20   1000 tonnes    -14.24
21   1000 tonnes    -13.76
22   1000 tonnes    -13.25
23   1000 tonnes    -13.13
24   1000 tonnes    -11.20
25   1000 tonnes     -9.65
26   1000 tonnes     -9.19
27   1000 tonnes     -8.87
28   1000 tonnes     -6.37
29   1000 tonnes     -3.37
..           ...       ...
143  1000 tonnes     45.10
144  1000 tonnes     45.94
145  1000 tonnes     46.15
146  1000 tonnes     46.23
147  1000 tonnes     46.82
148  1000 tonnes     46.86
149  1000 tonnes     47.16
150  1000 tonnes     47.41
151  1000 tonnes     47.52
152  1000 tonnes     48.02
153  1000 tonnes     48.38
154  1000 tonnes     48.67
155  1000 tonnes     49.82
156  1000 tonnes     50.50
157  1000 tonnes     51.17
158  1000 tonnes     51.92
159  1000 tonnes     52.13
160  1000 tonnes     53.41
161  1000 tonnes     53.71
162  1000 tonnes     55.17
163  1000 tonnes     55.38
164  1000 tonnes     56.13
165  1000 tonnes     56.26
166  1000 tonnes     56.88
167  1000 tonnes     58.60
168  1000 tonnes     60.13
169  1000 tonnes     60.47
170  1000 tonnes     61.52
171  1000 tonnes     61.92
172  1000 tonnes     64.96

[173 rows x 2 columns]